asllib/aslspec/latex.ml

```diff
@@ -0,0 +1,267 @@
+let elem_name_to_math_macro elem_name = "\\" ^ remove_underscores elem_name
+let pp_comma = fun fmt () -> fprintf fmt ", "
+
+(** [substitute_spec_vars_by_latex_vars math_mode s vars] returns a string [s]
+    with every variable like [my_var] is substituted into [\texttt{my\_var}],
+    which makes the returned string suitable to typsetting with LaTeX. If
+    [math_mode] is true then the result is surrounded by [$$]. *)
+let substitute_spec_vars_by_latex_vars ~math_mode str vars =
+  let module StringMap = Map.Make (String) in
+  let substitutions_map =
+    List.fold_left
+      (fun acc_map var_str ->
+        let latex_var = spec_var_to_latex_var ~font_type:TextTT var_str in
+        let latex_var =
+          if math_mode then to_math_mode latex_var else latex_var
+        in
+        let template_var = spec_var_to_template_var var_str in
+        StringMap.add template_var latex_var acc_map)
+      StringMap.empty vars
+  in
+  let template_var_regexp = Str.regexp "{[a-zA-Z0-9_']+}" in
+  let blocks = Str.full_split template_var_regexp str in
+  List.fold_left
+    (fun acc block ->
+      match block with
+      | Str.Text txt -> acc ^ txt
+      | Str.Delim var -> (
+          match StringMap.find_opt var substitutions_map with
+          | Some latex_var -> acc ^ latex_var
+          | None -> acc ^ var))
+    "" blocks
+
+(** [pp_texthypertarget fmt target_str] renders a hypertarget for [target_str]
+    suitable for inclusion in LaTeX text mode, using the formatter [fmt]. *)
+let pp_texthypertarget fmt target_str =
+  fprintf fmt {|\texthypertarget{%s}|} target_str
+
+(** [pp_mathhypertarget fmt target_str] renders a hypertarget for [target_str]
+    suitable for inclusion in LaTeX math mode, using the formatter [fmt]. *)
+let pp_mathhypertarget fmt target_str =
+  fprintf fmt {|\mathhypertarget{%s}|} target_str
+
+(** A type enumerating different kinds of parentheses. *)
+type parenthesis = Parens | Braces | Brackets
+
+(** [pp_parenthesized parenthesis large pp_elem fmt elem] formats [elem] with
+    the formatter [fmt] using [pp_elem] and surrounds it with the specified
+    [parenthesis]. If [large] is true, then large versions of the parentheses
+    are used. *)
+let pp_parenthesized parenthesis large pp_elem fmt elem =
+  let left = if large then {|\left|} else "" in
+  let right = if large then {|\right|} else "" in
+  match parenthesis with
+  | Parens -> fprintf fmt "%s(%a%s)" left pp_elem elem right
+  | Braces -> fprintf fmt "%s\\{%a%s\\}" left pp_elem elem right
+  | Brackets -> fprintf fmt "%s[%a%s]" left pp_elem elem right
+
+(** [over_text fmt fmt_top top fmt_bottom bottom] renders
+    [\overtext{top}{bottom}] where [top] and [bottom] are formatted using
+    [fmt_top] and [fmt_bottom] respectively. *)
+let over_text fmt fmt_top top fmt_bottom bottom =
+  fprintf fmt {|\overtext{%a}{%a}|} fmt_top top fmt_bottom bottom
+
+(** [pp_field_name_to_latex field_name] formats [field_name] - a field name
```

Comment:
nit: comment doesn't match function name. Should the function be in the `pp_*` style for consistency?